python/cudf/cudf/utils/_numba.py

```diff
@@ -0,0 +1,196 @@
+run_major, run_minor = cuda.runtime.get_version()
+print(f'{drv_major} {drv_minor} {run_major} {run_minor}')
+"""
+
+
+def _get_versions():
+    """
+    This function is mostly vendored from ptxcompiler and is used
+    to check the system CUDA driver and runtime versions in its absence.
+    """
+    cp = subprocess.run([sys.executable, "-c", CMD], capture_output=True)
+    if cp.returncode:
+        return _NO_DRIVER
+
+    versions = [int(s) for s in cp.stdout.strip().split()]
+    driver_version = tuple(versions[:2])
+    runtime_version = tuple(versions[2:])
+
+    return driver_version, runtime_version
+
+
+def _get_best_ptx_file(archs, max_compute_capability):
+    """
+    Determine of the available PTX files which one is
+    the most recent up to and including the device compute capability.
+    """
+    filtered_archs = [x for x in archs if x[0] <= max_compute_capability]
+    if filtered_archs:
+        return max(filtered_archs, key=lambda x: x[0])
+    else:
+        return None
+
+
+def _get_ptx_file(path, prefix):
+    if "RAPIDS_NO_INITIALIZE" in os.environ:
+        # cc=60 ptx is always built
+        cc = int(os.environ.get("STRINGS_UDF_CC", "60"))
+    else:
+        from numba import cuda
+
+        dev = cuda.get_current_device()
+
+        # Load the highest compute capability file available that is less than
+        # the current device's.
+        cc = int("".join(str(x) for x in dev.compute_capability))
+    files = glob.glob(os.path.join(path, f"{prefix}*.ptx"))
+    if len(files) == 0:
+        raise RuntimeError(f"Missing PTX files for cc={cc}")
+    regular_sms = []
+
+    for f in files:
+        file_name = os.path.basename(f)
+        sm_number = file_name.rstrip(".ptx").lstrip(prefix)
+        if sm_number.endswith("a"):
+            processed_sm_number = int(sm_number.rstrip("a"))
+            if processed_sm_number == cc:
+                return f
+        else:
+            regular_sms.append((int(sm_number), f))
+
+    regular_result = None
+
+    if regular_sms:
+        regular_result = _get_best_ptx_file(regular_sms, cc)
+
+    if regular_result is None:
+        raise RuntimeError(
+            "This cuDF installation is missing the necessary PTX "
+            f"files that are <={cc}."
+        )
+    else:
+        return regular_result[1]
+
+
+def _setup_numba():
+    """
+    Configure the numba linker for use with cuDF. This consists of
+    potentially putting numba into enhanced compatibility mode
+    based on the user driver and runtime versions as well as the
+    version of the CUDA Toolkit used to build the PTX files shipped
+    with the user cuDF package.
+    """
+    try:
+        # By default, ptxcompiler will not be installed with CUDA 12
+        # packages. This is ok, because in this situation putting
+        # numba in enhanced compatibility mode is not necessary.
+        from ptxcompiler.patch import NO_DRIVER, safe_get_versions
+    except ModuleNotFoundError:
```

Comment:
To my knowledge, there's nothing prohibiting a user from installing `ptxcompiler` via pip or conda when using CUDA 12. If the absence of `ptxcompiler` is meant to be a proxy for using CUDA 12, we need a different strategy here.